opal/datatype/opal_copy_functions_heterogeneous.c

```diff
@@ -133,6 +164,469 @@ static inline void opal_dt_swap_long_double(void *to_p, const void *from_p, cons
+    (((arch) & OPAL_ARCH_LDEXPSIZEISxx) == OPAL_ARCH_LDEXPSIZEIS11)        \
+  )
+#define LDBL_IS_F80(arch) \
+  (                                                                        \
+    (((arch) & OPAL_ARCH_LDMANTDIGISxx) == OPAL_ARCH_LDMANTDIGIS64)        \
+    &&                                                                     \
+    (((arch) & OPAL_ARCH_LDEXPSIZEISxx) == OPAL_ARCH_LDEXPSIZEIS15)        \
+  )
+#define LDBL_IS_F128(arch) \
+  (                                                                        \
+    (((arch) & OPAL_ARCH_LDMANTDIGISxx) == OPAL_ARCH_LDMANTDIGIS113)       \
+    &&                                                                     \
+    (((arch) & OPAL_ARCH_LDEXPSIZEISxx) == OPAL_ARCH_LDEXPSIZEIS15)        \
+  )
+#define LDBL_INFO_MASK (OPAL_ARCH_LDMANTDIGISxx | OPAL_ARCH_LDEXPSIZEISxx)
+
+#ifdef HAVE___FLOAT128
+/*
+ *  I'm not sure about the portability of alignof() so I'm handling things
+ *  like the possibility of sizeof(long double) == 12 in a slower way.  The
+ *  alignment requirement in that case would be 4 (largest power of 2 that
+ *  divides into the sizeof).
+ *
+ *  And saving it static to just compute it once without running a loop
+ *  every call.
+ */
+static inline
+size_t
+alignment_of_long_double() {
+    static size_t val = 0;
+
+    if (val == 0) {
+        val = 1;
+        while (sizeof(long double) % (val*2) == 0) {
+            val *= 2;
+        }
+    }
+    return val;
+}
+#endif
+
+// ldbl_to_f128 (copies a long double(from_arch format) to a float128(local_endian))
+static inline
+void
+ldbl_to_f128(unsigned char *f128_buf_to, const unsigned char *ldbl_buf_from, ssize_t count, int from_arch, ptrdiff_t from_extent)
+{
+#ifdef HAVE___FLOAT128
+    int ldbl_is_aligned;
+
+    ldbl_is_aligned = 1;
+    int alignment_mask = alignment_of_long_double() - 1;
+    if ((uintptr_t)ldbl_buf_from & alignment_mask) {
+        ldbl_is_aligned = 0;
+    }
+    if ((uintptr_t)from_extent & alignment_mask) {
+        ldbl_is_aligned = 0;
+    }
+
+    int f128_is_aligned;
+    f128_is_aligned = 1;
+    if ((uintptr_t)f128_buf_to & 0xF) {
+        f128_is_aligned = 0;
+    }
+
+    do {
+        if (ldbl_is_aligned && f128_is_aligned) {
+            *(__float128*)f128_buf_to = *(long double*)ldbl_buf_from;
+        } else {
+            __float128 f128;
+            long double ldbl;
+            memcpy(&ldbl, ldbl_buf_from, sizeof(ldbl));
+            f128 = ldbl;
+            memcpy(f128_buf_to, &f128, sizeof(f128));
+        }
+
+        ldbl_buf_from += from_extent;
+        f128_buf_to += sizeof(__float128);
+        count--;
+    } while (count > 0);
+#else
+    if (LDBL_IS_F64(from_arch)) {
+        f64_to_f128(f128_buf_to, ldbl_buf_from, count, from_extent);
+    } else if (LDBL_IS_F80(from_arch)) {
+        f80_to_f128(f128_buf_to, ldbl_buf_from, count, from_extent);
+    } else {
+/*
+ *  This could be an error condiiton, eg we're trying to process a
```

Comment:
typo in condition.

Reply:
I don't see the typo, what are you seeing wrong there?

Reply:
two swapped letters in `condiiton`

Reply:
Oh right, thx